src/workerd/api/worker-rpc.c++

```diff
@@ -148,64 +148,182 @@ private:
+  path.add(name);
+  return result;
+}
+
+template <typename FillOpFunc>
+JsRpcProperty::PromiseAndPipleine JsRpcProperty::callImpl(jsg::Lock& js, FillOpFunc&& fillOpFunc) {
+  // Note: We used to enforce that RPC methods had to be called with the correct `this`. That is,
+  // we prevented people from doing:
+  //
+  //   let obj = {foo: someRpcStub.foo};
+  //   obj.foo();
+  //
+  // This would throw "Illegal invocation", as is the norm when pulling methods of a native object.
+  // That worked as long as RPC methods were implemented as `jsg::Function`. However, when we
+  // switched to RPC methods being implemented as callable objects (JsRpcProperty), this became
+  // impossible, because V8's SetCallAsFunctionHandler() arranges that `this` is bound to the
+  // callable object itself, regardless of how it was invoked. So now we cannot detect the
+  // situation above, because V8 never tells us about `obj` at all.
+  //
+  // Oh well. It's not a big deal. Just annoying that we have to forever support tearing RPC
+  // methods off their source object, even if we change implementations to something where that's
+  // less convenient.
+
+  kj::Vector<kj::StringPtr> path;
+  auto client = parent->getClientForOneCall(js, path);
+
   auto& ioContext = IoContext::current();
 
   auto builder = client.callRequest();
-  builder.setMethodName(name);
 
-  kj::Vector<jsg::JsValue> argv(args.Length());
-  for (int n = 0; n < args.Length(); n++) {
-    argv.add(jsg::JsValue(args[n]));
+  if (path.empty()) {
+    builder.setMethodName(name);
+  } else {
+    auto pathBuilder = builder.initMethodPath(path.size() + 1);
+    for (auto i: kj::indices(path)) {
+      pathBuilder.set(i, path[i]);
+    }
+    pathBuilder.set(path.size(), name);
   }
 
-  // If we have arguments, serialize them.
-  // Note that we may fail to serialize some element, in which case this will throw back to JS.
-  if (argv.size() > 0) {
-    serializeJsValue(js, js.arr(argv.asPtr()), [&](capnp::MessageSize hint) {
-      // TODO(perf): Actually use the size hint.
-      return builder.initArgs();
-    });
-  }
+  fillOpFunc(builder.getOperation());
 
   auto callResult = builder.send();
 
-  return ioContext.awaitIo(js, kj::mv(callResult),
-      [](jsg::Lock& js, auto result) -> jsg::Value {
-    return jsg::Value(js.v8Isolate, deserializeJsValue(js, result.getResult()));
+  return {
+    .promise = jsg::JsPromise(js.wrapSimplePromise(ioContext.awaitIo(js, kj::mv(callResult),
+        [](jsg::Lock& js, auto result) -> jsg::Value {
+      return jsg::Value(js.v8Isolate, deserializeJsValue(js, result.getResult()));
+    }))),
+    .pipeline = kj::mv(callResult),
+  };
+}
+
+jsg::Ref<JsRpcPromise> JsRpcProperty::call(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  jsg::Lock& js = jsg::Lock::from(args.GetIsolate());
+
+  auto [promise, pipeline] = callImpl(js, [&](rpc::JsRpcTarget::CallParams::Operation::Builder op) {
+    kj::Vector<jsg::JsValue> argv(args.Length());
```

Comment:
non-blocking nit.. would KJ_STACK_ARRAY be ok here?

Reply:
Looks like it, yeah. (Not new code in this PR but I'll go ahead and improve it.)

Reply:
Actually it doesn't work since JsValue doesn't have a default constructor. Will have to revisit later.